### DISCUSSION TITLE
Dynamically create OPI content from YAMCS queries

### DISCUSSION THREAD

**Discussion by mbenson1** [yamcs/yamcs-studio#103] [Q&A]
I created (and retired) a completely web based frontend to YAMCS.  I'm migrating back to YAMCS Studio now.  One of the features of the frontend I created is it automatically created command argument popups.  I could specify none, some, or all arguments in the command button metadata.  The frontend would query the server for command definitions.  Then it would match up and validate the arguments supplied in the command button metadata, if any.  If there were still any arguments left undefined in the metadata, it would generate a popup and query the user to supply the remaining arguments.  I work with systems that have thousands of commands, most of which have arguments.  Hand writing a set of OPI's just for command arguments can take a lot of time and is error prone.

1.  Is it possible to query the YAMCS server for command definitions, including argument definitions?  If so, can somebody post an example?
2.  Is it possible to automatically run a script when an OPI is opened?  If so, can somebody post an example?
3.  Is it possible to pass arguments defined at build time in metadata, to an OPI at runtime?  I see the "Macros" property in the "Set Actions" property for buttons.  Is that how you pass key/value pairs to the OPI?  If so, could somebody post an example of how to read the macros inside a script?
4.  Is it possible to add widgets to a window at runtime?  I don't mean generating an OPI file at build time, though that is a fallback option.  I mean creating an OPI that is basically empty but has a large script that executes automatically when the window is opened.  The script would dynamically add the widgets.  When the window is closed, everything is deleted.  When the window is opened again, the script reruns again and adds the widgets back again.  Assuming the above capability (3) is possible, a single dynamic OPI could be used to for all command buttons.  The caller would just pass the full ID of the command to the creation script.  If this is possible, can somebody post an example?

If all 4 of these capabilities are possible, I think dynamic creation of OPI content from the YAMCS database is possible.  I probably wouldn't use this method universally, but it would at least be nice to have this as an option early in development.  Either as a temporary method until the user has time to hand generate better looking more user friendly displays.  Or, when this produces displays that meet production requirements, as a permanent alternative to hand generating every command argument OPI.

**Comment by fqqb** [comment]
Hi, I see where you are going, but for such deep integration, I also think you will need to reverse engineer from existing code (the widgeting is a forked from CS-Studio), and adapt if/where required. I would need to do exactly the same before I could answer your questions properly.

So quick take:

1/ Not yet. But it could be added in https://github.com/yamcs/yamcs-studio/blob/master/bundles/org.yamcs.studio.css.script/src/main/java/org/yamcs/studio/script/Yamcs.java. You have access to the applicable YamcsClient there.

2/ Attach a script to the display itself and have that trigger with a `loc://bla(1)` PV? The (1) makes it initialize with the value 1.

3/ I suppose you've found `ScriptUtil.openOPI` already? It has a macros argument that you can make using `DataUtil.createMacrosInput`. Inside script, `widget.getMacroValue("...");` seems to do something for me. As does `display.getMacroValue("...");`

4/ `WidgetUtil.createWidgetModel` appears to do that. Probably in combination with the `display` object that is present in a script context too (set by https://github.com/yamcs/yamcs-studio/blob/master/bundles/org.csstudio.opibuilder/src/org/csstudio/opibuilder/script/JavaScriptStore.java#L39)
